[Bug] [kafka-source] Multiple KafkaSource failed with Only one StreamingContext may be started in this JVM.

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-seatunnel/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.


### What happened

I add seatunnel as maven dependencies to start my service, when I create two kafka-source task, I found only one task works, another one failed.

### SeaTunnel Version

V1.2.2

### SeaTunnel Config

```conf
env {
  spark.app.name = "SeaTunnel"
  spark.executor.instances = 2
  spark.executor.cores = 1
  spark.executor.memory = "1g"
  spark.master = "local"
  spark.serializer = "org.apache.spark.serializer.KryoSerializer"
  spark.yarn.archive = "hdfs:///app-logs/spark-2.4.4-1.0.2.zip"
  spark.proxy-user = "spark"
  spark.driver.allowMultipleContexts = true
}

source {
  KafkaStream {
    topics = "test1"
    consumer.bootstrap.servers = "xxx.xxx.xxx.xxx:9092"
    consumer.group.id = "consumer0001"
    result_table_name = "test"
  }
}

transform {
  Sql {
    sql = "SELECT * FROM test"
  }
}

sink {
   Console {
   }
}
```


### Running Command

```shell
import org.apache.seatunnel.common.config.DeployMode;
import org.apache.seatunnel.core.base.Seatunnel;
import org.apache.seatunnel.core.base.command.Command;
import org.apache.seatunnel.core.spark.args.SparkCommandArgs;
import org.apache.seatunnel.core.spark.command.SparkCommandBuilder;

import java.io.FileNotFoundException;
import java.net.URISyntaxException;
import java.net.URL;
import java.nio.file.Paths;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class SparkEngineExample {

    static class SparkTask implements Runnable {

        private String confFile;

        public SparkTask(String confFile) {
            this.confFile = confFile;
        }

        public String getTestConfigFile(String configFile) throws URISyntaxException, FileNotFoundException {
            URL resource = SparkEngineExample.class.getResource(configFile);
            if (resource == null) {
                throw new FileNotFoundException("Could not find config file: " + configFile);
            }
            return Paths.get(resource.toURI()).toString();
        }

        @Override
        public void run() {
            String configFile = null;
            try {
                configFile = getTestConfigFile(this.confFile);
            } catch (URISyntaxException e) {
                e.printStackTrace();
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }
            SparkCommandArgs sparkArgs = new SparkCommandArgs();
            sparkArgs.setConfigFile(configFile);
            sparkArgs.setCheckConfig(false);
            sparkArgs.setVariables(null);
            sparkArgs.setDeployMode(DeployMode.CLIENT);
            Command<SparkCommandArgs> sparkCommand = new SparkCommandBuilder().buildCommand(sparkArgs);
            Seatunnel.run(sparkCommand);
        }
    }

    public static void main(String[] args) {
        ExecutorService service = Executors.newCachedThreadPool();
        service.execute(new SparkTask("/examples/test1.conf"));
        service.execute(new SparkTask("/examples/test2.conf"));
        service.shutdown();
    }
}
```


### Error Exception

```log
22/08/31 18:53:41 ERROR Seatunnel: 

===============================================================================


22/08/31 18:53:41 ERROR Seatunnel: Fatal Error, 

22/08/31 18:53:41 ERROR Seatunnel: Please submit bug report in https://github.com/apache/incubator-seatunnel/issues

22/08/31 18:53:41 ERROR Seatunnel: Reason:Execute Spark task error 

22/08/31 18:53:41 ERROR Seatunnel: Exception StackTrace:java.lang.RuntimeException: Execute Spark task error
	at org.apache.seatunnel.core.spark.command.SparkTaskExecuteCommand.execute(SparkTaskExecuteCommand.java:79)
	at org.apache.seatunnel.core.base.Seatunnel.run(Seatunnel.java:39)
	at com.somnus.SparkEngineExample$SparkTask.run(SparkEngineExample.java:67)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: Only one StreamingContext may be started in this JVM. Currently running StreamingContext was started atorg.apache.spark.streaming.StreamingContext.start(StreamingContext.scala:569)
org.apache.seatunnel.spark.stream.SparkStreamingExecution.start(SparkStreamingExecution.scala:68)
org.apache.seatunnel.core.spark.command.SparkTaskExecuteCommand.execute(SparkTaskExecuteCommand.java:76)
org.apache.seatunnel.core.base.Seatunnel.run(Seatunnel.java:39)
com.somnus.SparkEngineExample$SparkTask.run(SparkEngineExample.java:67)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
java.lang.Thread.run(Thread.java:748)
	at org.apache.spark.streaming.StreamingContext$.org$apache$spark$streaming$StreamingContext$$assertNoOtherContextIsActive(StreamingContext.scala:738)
	at org.apache.spark.streaming.StreamingContext.start(StreamingContext.scala:571)
	at org.apache.seatunnel.spark.stream.SparkStreamingExecution.start(SparkStreamingExecution.scala:68)
	at org.apache.seatunnel.core.spark.command.SparkTaskExecuteCommand.execute(SparkTaskExecuteCommand.java:76)
	... 5 more
 
22/08/31 18:53:41 INFO ContextHandler: Started o.s.j.s.ServletContextHandler@16c74877{/streaming,null,AVAILABLE,@Spark}
```


### Flink or Spark Version

Spark2.4.4

### Java or Scala Version

Scala2.11.12

### Screenshots

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaConsumerThread.java
+++ b/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaConsumerThread.java
@@ -1,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.seatunnel.connectors.seatunnel.kafka.source;
+
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+
+import java.util.Properties;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+public class KafkaConsumerThread implements Runnable {
+
+    private final KafkaConsumer<byte[], byte[]> consumer;
+    private static final String CLIENT_ID_PREFIX = "seatunnel";
+    private final ConsumerMetadata metadata;
+
+    private final LinkedBlockingQueue<Consumer<KafkaConsumer<byte[], byte[]>>> tasks;
+
+    public KafkaConsumerThread(ConsumerMetadata metadata) {
+        this.metadata = metadata;
+        this.tasks = new LinkedBlockingQueue<>();
+        this.consumer = initConsumer(this.metadata.getBootstrapServers(), this.metadata.getConsumerGroup(),
+            this.metadata.getProperties(), !this.metadata.isCommitOnCheckpoint());
+    }
+
+    @Override
+    public void run() {
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                Consumer<KafkaConsumer<byte[], byte[]>> task = tasks.poll(1, TimeUnit.SECONDS);
+                if (task != null) {
+                    task.accept(consumer);
+                }
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    public LinkedBlockingQueue<Consumer<KafkaConsumer<byte[], byte[]>>> getTasks() {
+        return tasks;
+    }
+
+    private KafkaConsumer<byte[], byte[]> initConsumer(String bootstrapServer, String consumerGroup,
+                                                       Properties properties, boolean autoCommit) {
+        Properties props = new Properties();
+        properties.forEach((key, value) -> props.setProperty(String.valueOf(key), String.valueOf(value)));
+        props.setProperty(ConsumerConfig.GROUP_ID_CONFIG, consumerGroup);
+        props.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServer);
+        props.setProperty(ConsumerConfig.CLIENT_ID_CONFIG, CLIENT_ID_PREFIX + "-enumerator-consumer-" + this.hashCode());
+
+        props.setProperty(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+            ByteArrayDeserializer.class.getName());
+        props.setProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+            ByteArrayDeserializer.class.getName());
+        props.setProperty(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, String.valueOf(autoCommit));
+
+        // Disable auto create topics feature
+        props.setProperty(ConsumerConfig.ALLOW_AUTO_CREATE_TOPICS_CONFIG, "false");
+        return new KafkaConsumer<>(props);
+    }
+}

--- a/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceReader.java
+++ b/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceReader.java
@@ -24,12 +24,11 @@ import org.apache.seatunnel.api.table.type.SeaTunnelRow;
 import org.apache.seatunnel.api.table.type.SeaTunnelRowType;
 
 import com.google.common.collect.Maps;
-import org.apache.kafka.clients.consumer.ConsumerConfig;
+import com.google.common.collect.Sets;
+import lombok.extern.slf4j.Slf4j;
 import org.apache.kafka.clients.consumer.ConsumerRecord;
 import org.apache.kafka.clients.consumer.ConsumerRecords;
-import org.apache.kafka.clients.consumer.KafkaConsumer;
 import org.apache.kafka.common.TopicPartition;
-import org.apache.kafka.common.serialization.ByteArrayDeserializer;
 import org.apache.kafka.common.serialization.StringDeserializer;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -37,30 +36,32 @@ import org.slf4j.LoggerFactory;
 import java.io.IOException;
 import java.time.Duration;
 import java.util.ArrayList;
-import java.util.Collection;
-import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
-import java.util.Properties;
 import java.util.Set;
-import java.util.stream.Collectors;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 
+@Slf4j
 public class KafkaSourceReader implements SourceReader<SeaTunnelRow, KafkaSourceSplit> {
 
     private static final long THREAD_WAIT_TIME = 500L;
     private static final long POLL_TIMEOUT = 10000L;
-    private static final String CLIENT_ID_PREFIX = "seatunnel";
+
     private static final Logger LOGGER = LoggerFactory.getLogger(KafkaSourceReader.class);
 
     private final SourceReader.Context context;
-    private KafkaConsumer<byte[], byte[]> consumer;
     private final ConsumerMetadata metadata;
     private final Set<KafkaSourceSplit> sourceSplits;
-    private final Map<TopicPartition, Long> endOffset;
+    private final ConcurrentMap<TopicPartition, KafkaSourceSplit> sourceSplitMap;
+    private final Map<KafkaSourceSplit, KafkaConsumerThread> consumerThreadMap;
+    private final ExecutorService executorService;
     // TODO support user custom type
     private SeaTunnelRowType typeInfo;
-    private volatile boolean isRunning;
 
     KafkaSourceReader(ConsumerMetadata metadata, SeaTunnelRowType typeInfo,
                       SourceReader.Context context) {
@@ -68,55 +69,84 @@ public class KafkaSourceReader implements SourceReader<SeaTunnelRow, KafkaSource
         this.context = context;
         this.typeInfo = typeInfo;
         this.sourceSplits = new HashSet<>();
-        this.endOffset = new HashMap<>();
+        this.consumerThreadMap = new ConcurrentHashMap<>();
+        this.sourceSplitMap = new ConcurrentHashMap<>();
+        this.executorService = Executors.newCachedThreadPool(
+            r -> new Thread(r, "Kafka Source Data Consumer"));
     }
 
     @Override
     public void open() {
-        this.consumer = initConsumer(this.metadata.getBootstrapServers(), this.metadata.getConsumerGroup(),
-                this.metadata.getProperties(), !this.metadata.isCommitOnCheckpoint());
-        isRunning = true;
     }
 
     @Override
     public void close() throws IOException {
-        isRunning = false;
-        if (consumer != null) {
-            consumer.close();
+        if (executorService != null) {
+            executorService.shutdownNow();
         }
     }
 
     @Override
     public void pollNext(Collector<SeaTunnelRow> output) throws Exception {
-        if (sourceSplits.isEmpty() || sourceSplits.size() != this.endOffset.size()) {
+        if (sourceSplitMap.isEmpty()) {
             Thread.sleep(THREAD_WAIT_TIME);
             return;
         }
-        Set<TopicPartition> partitions = convertToPartition(sourceSplits);
-        StringDeserializer stringDeserializer = new StringDeserializer();
-        stringDeserializer.configure(Maps.fromProperties(this.metadata.getProperties()), false);
-        consumer.assign(partitions);
-        while (isRunning) {
-            ConsumerRecords<byte[], byte[]> records = consumer.poll(Duration.ofMillis(POLL_TIMEOUT));
-            for (TopicPartition partition : partitions) {
-                for (ConsumerRecord<byte[], byte[]> record : records.records(partition)) {
+        sourceSplits.forEach(sourceSplit -> consumerThreadMap.computeIfAbsent(sourceSplit, s -> {
+            KafkaConsumerThread thread = new KafkaConsumerThread(metadata);
+            executorService.submit(thread);
+            return thread;
+        }));
+        sourceSplits.forEach(sourceSplit -> {
+            CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+            try {
+                consumerThreadMap.get(sourceSplit).getTasks().put(consumer -> {
+                    try {
+                        Set<TopicPartition> partitions = Sets.newHashSet(sourceSplit.getTopicPartition());
+                        StringDeserializer stringDeserializer = new StringDeserializer();
+                        stringDeserializer.configure(Maps.fromProperties(this.metadata.getProperties()), false);
+                        consumer.assign(partitions);
+                        if (sourceSplit.getStartOffset() >= 0) {
+                            consumer.seek(sourceSplit.getTopicPartition(), sourceSplit.getStartOffset());
+                        }
+                        ConsumerRecords<byte[], byte[]> records = consumer.poll(Duration.ofMillis(POLL_TIMEOUT));
+                        for (TopicPartition partition : partitions) {
+                            List<ConsumerRecord<byte[], byte[]>> recordList = records.records(partition);
+                            for (ConsumerRecord<byte[], byte[]> record : recordList) {
 
-                    String v = stringDeserializer.deserialize(partition.topic(), record.value());
-                    String t = partition.topic();
-                    output.collect(new SeaTunnelRow(new Object[]{t, v}));
+                                String v = stringDeserializer.deserialize(partition.topic(), record.value());
+                                String t = partition.topic();
+                                output.collect(new SeaTunnelRow(new Object[]{t, v}));
 
-                    if (Boundedness.BOUNDED.equals(context.getBoundedness()) &&
-                            record.offset() >= endOffset.get(partition)) {
-                        break;
+                                if (Boundedness.BOUNDED.equals(context.getBoundedness()) &&
+                                    record.offset() >= sourceSplit.getEndOffset()) {
+                                    break;
+                                }
+                            }
+                            long lastOffset = -1;
+                            if (!recordList.isEmpty()) {
+                                lastOffset = recordList.get(recordList.size() - 1).offset();
+                                sourceSplit.setStartOffset(lastOffset + 1);
+                            }
+
+                            if (lastOffset >= sourceSplit.getEndOffset()) {
+                                sourceSplit.setEndOffset(lastOffset);
+                            }
+                        }
+                    } catch (Exception e) {
+                        completableFuture.completeExceptionally(e);
                     }
-                }
+                    completableFuture.complete(null);
+                });
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
             }
+            completableFuture.join();
+        });
 
-            if (Boundedness.BOUNDED.equals(context.getBoundedness())) {
-                // signal to the source that we have reached the end of the data.
-                context.signalNoMoreElement();
-                break;
-            }
+        if (Boundedness.BOUNDED.equals(context.getBoundedness())) {
+            // signal to the source that we have reached the end of the data.
+            context.signalNoMoreElement();
         }
     }
 
@@ -128,8 +158,8 @@ public class KafkaSourceReader implements SourceReader<SeaTunnelRow, KafkaSource
     @Override
     public void addSplits(List<KafkaSourceSplit> splits) {
         sourceSplits.addAll(splits);
-        sourceSplits.forEach(partition -> {
-            endOffset.put(partition.getTopicPartition(), partition.getEndOffset());
+        sourceSplits.forEach(split -> {
+            sourceSplitMap.put(split.getTopicPartition(), split);
         });
     }
 
@@ -140,32 +170,17 @@ public class KafkaSourceReader implements SourceReader<SeaTunnelRow, KafkaSource
 
     @Override
     public void notifyCheckpointComplete(long checkpointId) throws Exception {
-        if (this.metadata.isCommitOnCheckpoint()) {
-            consumer.commitSync();
-        }
-    }
-
-    private KafkaConsumer<byte[], byte[]> initConsumer(String bootstrapServer, String consumerGroup,
-                                                       Properties properties, boolean autoCommit) {
-        Properties props = new Properties();
-        properties.forEach((key, value) -> props.setProperty(String.valueOf(key), String.valueOf(value)));
-        props.setProperty(ConsumerConfig.GROUP_ID_CONFIG, consumerGroup);
-        props.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServer);
-        props.setProperty(ConsumerConfig.CLIENT_ID_CONFIG, CLIENT_ID_PREFIX + "-enumerator-consumer-" + this.hashCode());
-
-        props.setProperty(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
-                ByteArrayDeserializer.class.getName());
-        props.setProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
-                ByteArrayDeserializer.class.getName());
-        props.setProperty(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, String.valueOf(autoCommit));
-
-        // Disable auto create topics feature
-        props.setProperty(ConsumerConfig.ALLOW_AUTO_CREATE_TOPICS_CONFIG, "false");
-        return new KafkaConsumer<>(props);
-    }
-
-    private Set<TopicPartition> convertToPartition(Collection<KafkaSourceSplit> sourceSplits) {
-        return sourceSplits.stream().map(KafkaSourceSplit::getTopicPartition).collect(Collectors.toSet());
+        consumerThreadMap.forEach((split, consumerThread) -> {
+            try {
+                consumerThread.getTasks().put(consumer -> {
+                    if (this.metadata.isCommitOnCheckpoint()) {
+                        consumer.commitSync();
+                    }
+                });
+            } catch (InterruptedException e) {
+                log.error("commit offset to kafka failed", e);
+            }
+        });
     }
 
 }

--- a/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceReader.java
+++ b/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceReader.java
@@ -28,6 +28,7 @@ import com.google.common.collect.Sets;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.kafka.clients.consumer.ConsumerRecord;
 import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
 import org.apache.kafka.common.TopicPartition;
 import org.apache.kafka.common.serialization.StringDeserializer;
 import org.slf4j.Logger;
@@ -36,6 +37,7 @@ import org.slf4j.LoggerFactory;
 import java.io.IOException;
 import java.time.Duration;
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
@@ -174,7 +176,9 @@ public class KafkaSourceReader implements SourceReader<SeaTunnelRow, KafkaSource
             try {
                 consumerThread.getTasks().put(consumer -> {
                     if (this.metadata.isCommitOnCheckpoint()) {
-                        consumer.commitSync();
+                        Map<TopicPartition, OffsetAndMetadata> offsets = new HashMap<>();
+                        offsets.put(split.getTopicPartition(), new OffsetAndMetadata(split.getEndOffset()));
+                        consumer.commitSync(offsets);
                     }
                 });
             } catch (InterruptedException e) {

--- a/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceReader.java
+++ b/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceReader.java
@@ -36,7 +36,6 @@ import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
 import java.time.Duration;
-import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
@@ -47,6 +46,7 @@ import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
 
 @Slf4j
 public class KafkaSourceReader implements SourceReader<SeaTunnelRow, KafkaSourceSplit> {
@@ -59,8 +59,9 @@ public class KafkaSourceReader implements SourceReader<SeaTunnelRow, KafkaSource
     private final SourceReader.Context context;
     private final ConsumerMetadata metadata;
     private final Set<KafkaSourceSplit> sourceSplits;
+    private final Map<Long, Map<TopicPartition, Long>> checkpointOffsetMap;
     private final ConcurrentMap<TopicPartition, KafkaSourceSplit> sourceSplitMap;
-    private final Map<KafkaSourceSplit, KafkaConsumerThread> consumerThreadMap;
+    private final Map<TopicPartition, KafkaConsumerThread> consumerThreadMap;
     private final ExecutorService executorService;
     // TODO support user custom type
     private SeaTunnelRowType typeInfo;
@@ -73,6 +74,7 @@ public class KafkaSourceReader implements SourceReader<SeaTunnelRow, KafkaSource
         this.sourceSplits = new HashSet<>();
         this.consumerThreadMap = new ConcurrentHashMap<>();
         this.sourceSplitMap = new ConcurrentHashMap<>();
+        this.checkpointOffsetMap = new ConcurrentHashMap<>();
         this.executorService = Executors.newCachedThreadPool(
             r -> new Thread(r, "Kafka Source Data Consumer"));
     }
@@ -94,7 +96,7 @@ public class KafkaSourceReader implements SourceReader<SeaTunnelRow, KafkaSource
             Thread.sleep(THREAD_WAIT_TIME);
             return;
         }
-        sourceSplits.forEach(sourceSplit -> consumerThreadMap.computeIfAbsent(sourceSplit, s -> {
+        sourceSplits.forEach(sourceSplit -> consumerThreadMap.computeIfAbsent(sourceSplit.getTopicPartition(), s -> {
             KafkaConsumerThread thread = new KafkaConsumerThread(metadata);
             executorService.submit(thread);
             return thread;
@@ -102,7 +104,7 @@ public class KafkaSourceReader implements SourceReader<SeaTunnelRow, KafkaSource
         sourceSplits.forEach(sourceSplit -> {
             CompletableFuture<Void> completableFuture = new CompletableFuture<>();
             try {
-                consumerThreadMap.get(sourceSplit).getTasks().put(consumer -> {
+                consumerThreadMap.get(sourceSplit.getTopicPartition()).getTasks().put(consumer -> {
                     try {
                         Set<TopicPartition> partitions = Sets.newHashSet(sourceSplit.getTopicPartition());
                         StringDeserializer stringDeserializer = new StringDeserializer();
@@ -153,8 +155,10 @@ public class KafkaSourceReader implements SourceReader<SeaTunnelRow, KafkaSource
     }
 
     @Override
-    public List<KafkaSourceSplit> snapshotState(long checkpointId) throws Exception {
-        return new ArrayList<>(sourceSplits);
+    public List<KafkaSourceSplit> snapshotState(long checkpointId) {
+        checkpointOffsetMap.put(checkpointId, sourceSplits.stream()
+            .collect(Collectors.toMap(KafkaSourceSplit::getTopicPartition, KafkaSourceSplit::getEndOffset)));
+        return sourceSplits.stream().map(KafkaSourceSplit::copy).collect(Collectors.toList());
     }
 
     @Override
@@ -171,20 +175,25 @@ public class KafkaSourceReader implements SourceReader<SeaTunnelRow, KafkaSource
     }
 
     @Override
-    public void notifyCheckpointComplete(long checkpointId) throws Exception {
-        consumerThreadMap.forEach((split, consumerThread) -> {
-            try {
-                consumerThread.getTasks().put(consumer -> {
-                    if (this.metadata.isCommitOnCheckpoint()) {
-                        Map<TopicPartition, OffsetAndMetadata> offsets = new HashMap<>();
-                        offsets.put(split.getTopicPartition(), new OffsetAndMetadata(split.getEndOffset()));
-                        consumer.commitSync(offsets);
-                    }
-                });
-            } catch (InterruptedException e) {
-                log.error("commit offset to kafka failed", e);
-            }
-        });
+    public void notifyCheckpointComplete(long checkpointId) {
+        if (!checkpointOffsetMap.containsKey(checkpointId)) {
+            log.warn("checkpoint {} do not exist or have already been committed.", checkpointId);
+        } else {
+            checkpointOffsetMap.remove(checkpointId).forEach((topicPartition, offset) -> {
+                try {
+                    consumerThreadMap.get(topicPartition)
+                        .getTasks().put(consumer -> {
+                            if (this.metadata.isCommitOnCheckpoint()) {
+                                Map<TopicPartition, OffsetAndMetadata> offsets = new HashMap<>();
+                                offsets.put(topicPartition, new OffsetAndMetadata(offset));
+                                consumer.commitSync(offsets);
+                            }
+                        });
+                } catch (InterruptedException e) {
+                    log.error("commit offset to kafka failed", e);
+                }
+            });
+        }
     }
 
 }

--- a/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceSplit.java
+++ b/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceSplit.java
@@ -26,10 +26,19 @@ import java.util.Objects;
 public class KafkaSourceSplit implements SourceSplit {
 
     private TopicPartition topicPartition;
+    private long startOffset = -1L;
     private long endOffset = -1L;
 
     public KafkaSourceSplit(TopicPartition topicPartition) {
         this.topicPartition = topicPartition;
+    }
+
+    public long getStartOffset() {
+        return startOffset;
+    }
+
+    public void setStartOffset(long startOffset) {
+        this.startOffset = startOffset;
     }
 
     public long getEndOffset() {
@@ -67,6 +76,6 @@ public class KafkaSourceSplit implements SourceSplit {
 
     @Override
     public int hashCode() {
-        return Objects.hash(topicPartition, endOffset);
+        return Objects.hash(topicPartition);
     }
 }

--- a/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceSplit.java
+++ b/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceSplit.java
@@ -33,6 +33,12 @@ public class KafkaSourceSplit implements SourceSplit {
         this.topicPartition = topicPartition;
     }
 
+    public KafkaSourceSplit(TopicPartition topicPartition, long startOffset, long endOffset) {
+        this.topicPartition = topicPartition;
+        this.startOffset = startOffset;
+        this.endOffset = endOffset;
+    }
+
     public long getStartOffset() {
         return startOffset;
     }
@@ -77,5 +83,9 @@ public class KafkaSourceSplit implements SourceSplit {
     @Override
     public int hashCode() {
         return Objects.hash(topicPartition);
+    }
+
+    public KafkaSourceSplit copy() {
+        return new KafkaSourceSplit(this.topicPartition, this.getStartOffset(), this.getEndOffset());
     }
 }

--- a/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceSplitEnumerator.java
+++ b/seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceSplitEnumerator.java
@@ -23,6 +23,7 @@ import org.apache.seatunnel.connectors.seatunnel.kafka.state.KafkaSourceState;
 
 import lombok.extern.slf4j.Slf4j;
 import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.ListOffsetsResult;
 import org.apache.kafka.clients.admin.OffsetSpec;
 import org.apache.kafka.clients.consumer.ConsumerConfig;
 import org.apache.kafka.common.TopicPartition;
@@ -51,29 +52,32 @@ public class KafkaSourceSplitEnumerator implements SourceSplitEnumerator<KafkaSo
     private AdminClient adminClient;
 
     private Set<KafkaSourceSplit> pendingSplit;
-    private Set<KafkaSourceSplit> assignedSplit;
+    private final Set<KafkaSourceSplit> assignedSplit;
 
     KafkaSourceSplitEnumerator(ConsumerMetadata metadata, Context<KafkaSourceSplit> context) {
         this.metadata = metadata;
         this.context = context;
+        this.assignedSplit = new HashSet<>();
+        this.pendingSplit = new HashSet<>();
     }
 
     KafkaSourceSplitEnumerator(ConsumerMetadata metadata, Context<KafkaSourceSplit> context,
                                KafkaSourceState sourceState) {
         this(metadata, context);
-        this.assignedSplit = sourceState.getAssignedSplit();
     }
 
     @Override
     public void open() {
         this.adminClient = initAdminClient(this.metadata.getProperties());
-        this.assignedSplit = new HashSet<>();
-        this.pendingSplit = new HashSet<>();
     }
 
     @Override
     public void run() throws ExecutionException, InterruptedException {
-        pendingSplit = getTopicInfo();
+        getTopicInfo().forEach(split -> {
+            if (!assignedSplit.contains(split)) {
+                pendingSplit.add(split);
+            }
+        });
         assignSplit();
     }
 
@@ -87,8 +91,22 @@ public class KafkaSourceSplitEnumerator implements SourceSplitEnumerator<KafkaSo
     @Override
     public void addSplitsBack(List<KafkaSourceSplit> splits, int subtaskId) {
         if (!splits.isEmpty()) {
-            pendingSplit.addAll(splits);
+            pendingSplit.addAll(convertToNextSplit(splits));
             assignSplit();
+        }
+    }
+
+    private Collection<? extends KafkaSourceSplit> convertToNextSplit(List<KafkaSourceSplit> splits) {
+        try {
+            Map<TopicPartition, ListOffsetsResult.ListOffsetsResultInfo> listOffsets =
+                getKafkaPartitionLatestOffset(splits.stream().map(KafkaSourceSplit::getTopicPartition).collect(Collectors.toList()));
+            splits.forEach(split -> {
+                split.setStartOffset(split.getEndOffset() + 1);
+                split.setEndOffset(listOffsets.get(split.getTopicPartition()).offset());
+            });
+            return splits;
+        } catch (Exception e) {
+            throw new RuntimeException(e);
         }
     }
 
@@ -131,30 +149,39 @@ public class KafkaSourceSplitEnumerator implements SourceSplitEnumerator<KafkaSo
         if (this.metadata.isPattern()) {
             Pattern pattern = Pattern.compile(this.metadata.getTopic());
             topics = this.adminClient.listTopics().names().get().stream()
-                    .filter(t -> pattern.matcher(t).matches()).collect(Collectors.toSet());
+                .filter(t -> pattern.matcher(t).matches()).collect(Collectors.toSet());
         } else {
             topics = Arrays.asList(this.metadata.getTopic().split(","));
         }
         log.info("Discovered topics: {}", topics);
 
         Collection<TopicPartition> partitions =
-                adminClient.describeTopics(topics).all().get().values().stream().flatMap(t -> t.partitions().stream()
-                        .map(p -> new TopicPartition(t.name(), p.partition()))).collect(Collectors.toSet());
+            adminClient.describeTopics(topics).all().get().values().stream().flatMap(t -> t.partitions().stream()
+                .map(p -> new TopicPartition(t.name(), p.partition()))).collect(Collectors.toSet());
+        return getKafkaPartitionLatestOffset(partitions).entrySet().stream().map(partition -> {
+            KafkaSourceSplit split = new KafkaSourceSplit(partition.getKey());
+            split.setEndOffset(partition.getValue().offset());
+            return split;
+        }).collect(Collectors.toSet());
+    }
+
+    private Map<TopicPartition, ListOffsetsResult.ListOffsetsResultInfo> getKafkaPartitionLatestOffset(Collection<TopicPartition> partitions) throws InterruptedException, ExecutionException {
         return adminClient.listOffsets(partitions.stream().collect(Collectors.toMap(p -> p, p -> OffsetSpec.latest())))
-                .all().get().entrySet().stream().map(partition -> {
-                    KafkaSourceSplit split = new KafkaSourceSplit(partition.getKey());
-                    split.setEndOffset(partition.getValue().offset());
-                    return split;
-                }).collect(Collectors.toSet());
+            .all().get();
     }
 
     private void assignSplit() {
         Map<Integer, List<KafkaSourceSplit>> readySplit = new HashMap<>(Common.COLLECTION_SIZE);
-        for (int taskID = 0;  taskID < context.currentParallelism(); taskID++) {
+        for (int taskID = 0; taskID < context.currentParallelism(); taskID++) {
             readySplit.computeIfAbsent(taskID, id -> new ArrayList<>());
         }
-        pendingSplit.forEach(s -> readySplit.get(getSplitOwner(s.getTopicPartition(), context.currentParallelism()))
-                .add(s));
+
+        pendingSplit.forEach(s -> {
+            if (!assignedSplit.contains(s)) {
+                readySplit.get(getSplitOwner(s.getTopicPartition(), context.currentParallelism()))
+                    .add(s);
+            }
+        });
 
         readySplit.forEach(context::assignSplit);
 

